### PR TITLE
Add HostMetrics receiver scaffolding code

### PR DESCRIPTION
**Pull request by james-bebbington**
_This PR is just for initial reviews/thoughts. I plan to get this to the point where it's actually doing something before creating a PR against the Otel repo._

**Description:**
Added initial scaffolding code for a new hostmetricsreceiver which will collect Host Metrics (CPU, Memory, Disk, etc) based on the operating system.

This receiver potentially needs quite a lot of configuration to specify which types of metrics should be collected, etc. This code expects the config to be in the format:
```
hostmetrics:
    scrape_interval: 1s
    cpu:
      reportpercpu: true
    memory:
    disk:
```
... where the absence of a metric type (e.g. `memory`) would mean we will not collect that metric.

I'm not sure if that's the cleanest way to have the config, but it was the easiest to implement. I could make it more dynamic in the future if needed.

Note I've used the word "collector" to describe the code that collects metrics of a certain type, but I will probably need to rename this as "collector" is a bit of an overloaded term in this repo.

TODOs:
* Add makefile
* Add README, etc
* Update to use new receiver interfaces
* Add a CircleCI build step for Windows, as the windows only code won't be compiled/tested as part of CI otherwise?
* Add e2e tests

### CODE REVIEW

File: testbed/tests/receivers.go [hunk 1]
```diff
@@ -178,3 +181,53 @@ func (cr *CarbonDataReceiver) GenConfigYAMLStr() string {
 func (cr *CarbonDataReceiver) ProtocolName() string {
 	return "carbon"
 }
+
+// HostMetricsReceiver implements Host Metrics receiver.
```
Comment (james-bebbington) [inline testbed/tests/receivers.go:185]: Some initial scaffolding for the e2e tests, but I haven't written them yet

File: receiver/hostmetricsreceiver/hostmetrics_receiver.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	configs := []struct {
+		config  hmcomponent.CollectorConfig
+		factory hmcomponent.CollectorFactory
+	}{
+		{config: config.CPUConfig, factory: &cpu.Factory{}},
+	}
+
+	collectors := make([]hmcomponent.Collector, 0)
+	for _, cfg := range configs {
+		if reflect.ValueOf(cfg.config).IsNil() {
```
Comment (james-bebbington) [inline receiver/hostmetricsreceiver/hostmetrics_receiver.go:71]: This is the code that ignores a metric type if it's not configured. Not 100% about the method for nil checking here (though reflection should be okay since this is only run on initialization)

Comment (nilebox) [reply receiver/hostmetricsreceiver/hostmetrics_receiver.go:71]: Not sure what are you trying to do here.
You've just defined `configs` inline above, with exactly 1 element. Why do you need to iterate over it?
Do you plan to add other types of configs there later?

P.S. the nil check via reflection is probably due to https://stackoverflow.com/questions/19761393/why-does-go-have-typed-nil

File: receiver/hostmetricsreceiver/collector/cpu/cpu_collector_others.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/hostmetricsreceiver/internal"
+)
+
+func (c *Collector) getPerCoreMetrics(reportPerCPU bool) []*metricspb.Metric {
+	return []*metricspb.Metric{
+		{
+			MetricDescriptor: MetricCPUSeconds,
+			Timeseries: []*metricspb.TimeSeries{
+				internal.GetDoubleTimeSeries(c.startTime, 0, LabelValueCPUUser, &metricspb.LabelValue{HasValue: false}),
+				internal.GetDoubleTimeSeries(c.startTime, 0, LabelValueCPUSystem, &metricspb.LabelValue{HasValue: false}),
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/collector/cpu/cpu_collector_others.go:31]: Do you want to define a "no value" constant for label in a similar way to  other label values?

File: receiver/hostmetricsreceiver/collector/cpu/cpu_constants.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+var MetricProcessCPUSeconds = &metricspb.MetricDescriptor{
+	Name:        "process/cpu/time",
+	Description: "CPU seconds for this process",
+	Unit:        "s",
+	Type:        metricspb.MetricDescriptor_CUMULATIVE_DOUBLE,
+	LabelKeys:   []*metricspb.LabelKey{{Key: "processname", Description: "Name of the process"}},
+}
+
+var (
+	LabelValueCPUUser   = &metricspb.LabelValue{Value: "user", HasValue: true}
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/collector/cpu/cpu_constants.go:51]: You may extract a function
```go
func labelValue(val string) *metricspb.LabelValue {
  return &metricspb.LabelValue{Value: val, HasValue: true}
}
```

File: receiver/hostmetricsreceiver/collector/cpu/factory_test.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+	"go.uber.org/zap"
+)
+
+func TestCreateMetricsCollector(t *testing.T) {
+	factory := &Factory{}
+	cfg := &Config{}
+
+	collector, err := factory.CreateMetricsCollector(zap.NewNop(), cfg)
+
+	assert.Nil(t, err)
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/collector/cpu/factory_test.go:30]:
```
require.NoError(t, err)
```

File: receiver/hostmetricsreceiver/component/collector.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+// limitations under the License.
+
+package component
+
+import (
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"go.uber.org/zap"
+)
+
+// Collector gathers metrics from the host machine and converts
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/component/collector.go:22]: maybe `Gatherer` then? 🙂

File: receiver/hostmetricsreceiver/config_test.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/open-telemetry/opentelemetry-collector/config/configmodels"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/hostmetricsreceiver/collector/cpu"
+)
+
+func TestLoadConfig(t *testing.T) {
+	factories, err := config.ExampleComponents()
+	assert.Nil(t, err)
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/config_test.go:32]:
```go
require.NoError(t, err)
```

File: receiver/hostmetricsreceiver/config_test.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+	require.NoError(t, err)
+	require.NotNil(t, cfg)
+
+	assert.Equal(t, len(cfg.Receivers), 2)
+
+	r0 := cfg.Receivers["hostmetrics"]
+	assert.Equal(t, r0, factory.CreateDefaultConfig())
+
+	r1 := cfg.Receivers["hostmetrics/customname"].(*Config)
+	assert.Equal(t, r1,
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/config_test.go:47]: Does this check work? You may need https://golang.org/pkg/reflect/#DeepEqual, and there is some other method to print the diff for testing IIRC.

Comment (nilebox) [reply receiver/hostmetricsreceiver/config_test.go:47]: Actually it's fine, this function is calling `ObjectsAreEqual` internally which invokes reflect.DeepEqual.
https://github.com/stretchr/testify/blob/master/assert/assertions.go#L343

File: receiver/hostmetricsreceiver/collector/cpu/cpu_collector_others.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/hostmetricsreceiver/internal"
+)
+
+func (c *Collector) getPerCoreMetrics(reportPerCPU bool) []*metricspb.Metric {
+	return []*metricspb.Metric{
+		{
+			MetricDescriptor: MetricCPUSeconds,
+			Timeseries: []*metricspb.TimeSeries{
+				internal.GetDoubleTimeSeries(c.startTime, 0, LabelValueCPUUser, &metricspb.LabelValue{HasValue: false}),
```
[the remark this answers is not in the document]
Comment (nilebox) [reply receiver/hostmetricsreceiver/collector/cpu/cpu_collector_others.go:30]: I suppose it's a stub implementation that will be properly implemented later.

File: receiver/hostmetricsreceiver/collector/cpu/cpu_constants.go [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+	LabelValueCPUSystem = &metricspb.LabelValue{Value: "system", HasValue: true}
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/collector/cpu/cpu_constants.go:52]: Should it be called `LabelValueCPUStateSystem`, given that it's a value for the "state" label key AFAICT?
Similar below.

File: receiver/hostmetricsreceiver/hostmetrics_receiver.go (hunk 2 above)
Comment (nilebox) [reply receiver/hostmetricsreceiver/hostmetrics_receiver.go:71]: Actually this seems like a bug, reflect will fail, see https://play.golang.org/p/DwGJQYip_ZL
So you need to replace it with
```go
if cfg.config != nil {
  ...
}
```

File: receiver/hostmetricsreceiver/component/collector.go (hunk 6 above)
Comment (rghetia) [reply receiver/hostmetricsreceiver/component/collector.go:22]: How about scraper (similar to prometheus)?

File: receiver/hostmetricsreceiver/hostmetrics_receiver.go (hunk 2 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/hostmetrics_receiver.go:71]: 1. Yea there are a bunch of metrics to collect so the idea is to add config for other "collectors" as per the PR description. I should probably express this as a list directly in the config so its clearer/tidier

i.e. change the definition of Config from:
```
type Config struct {
	configmodels.ReceiverSettings `mapstructure:",squash"`

	ScrapeInterval time.Duration `mapstructure:"scrape_interval"`
	CPUConfig      *cpu.Config   `mapstructure:"cpu"`
}
```
to something like :
```
type Config struct {
	configmodels.ReceiverSettings `mapstructure:",squash"`

	ScrapeInterval time.Duration `mapstructure:"scrape_interval"`
	CollectorConfigs      []*CollectorConfig   `mapstructure:"collectors"`
}
```

2. I think it's because of Go's distinction between the interface's concrete type being nil and value being nil:
https://medium.com/@mangatmodi/go-check-nil-interface-the-right-way-d142776edef1

Here's a small update to your playground to demonstrate how this works:
https://play.golang.org/p/BkWCOImyqpa

I guess it's potentially a sign of code smell that I need to perform this check, although it seems like a fairly common thing for people to run into. If I change to a list as per #1 this will go away anyway

File: receiver/hostmetricsreceiver/collector/cpu/cpu_collector_others.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/collector/cpu/cpu_collector_others.go:30]: Yea it was just a stub / dummy data

File: receiver/hostmetricsreceiver/collector/cpu/cpu_constants.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+import (
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+)
+
+// Host and process metric constants.
+
+var MetricCPUSeconds = &metricspb.MetricDescriptor{
+	Name:        "host/cpu/time",
+	Description: "Total CPU seconds broken down by different states",
+	Unit:        "s",
```
[the remark this answers is not in the document]
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/collector/cpu/cpu_constants.go:26]: Doesn't look like there's any currently defined - I could add some though the code comment for this property refers to http://unitsofmeasure.org/ucum.html which suggests a very broad set of possible values

File: receiver/hostmetricsreceiver/collector/cpu/cpu_collector_windows.go [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+				internal.GetDoubleTimeSeries(c.startTime, 0, LabelValueCPUIdle, &metricspb.LabelValue{HasValue: false}),
+				internal.GetDoubleTimeSeries(c.startTime, 0, LabelValueCPUNice, &metricspb.LabelValue{HasValue: false}),
+				internal.GetDoubleTimeSeries(c.startTime, 0, LabelValueCPUIOWait, &metricspb.LabelValue{HasValue: false}),
+			},
+		},
+		{
+			MetricDescriptor: MetricCPUUtilization,
+			Timeseries:       []*metricspb.TimeSeries{internal.GetDoubleTimeSeries(c.startTime, 0, &metricspb.LabelValue{HasValue: false})},
+		},
+	}
```
Comment (james-bebbington) [inline receiver/hostmetricsreceiver/collector/cpu/cpu_collector_windows.go:41]: @rghetia 

I was reading some of the (unfinished) metric spec documentation which refers to generating metric instruments via the Meter interface:
https://github.com/jmacd/opentelemetry-specification/blob/2a74584326655699d7c607dc885be0aba348cd20/specification/api-metrics.md#meter-interface

I see most of this is already implemented in the Go SDK (although the proto itself refers to Gauges and not counters so maybe that's not up-to-date yet?):
https://github.com/open-telemetry/opentelemetry-go/tree/master/api

I also see we have the idea of Observer metric type which should presumably be used for cpu utilization rather than Gauge metric (which is going to be deprecated).

So, basically I'm a bit confused as to if this is correct - is it okay to just generate metrics directly like this for now, or should I be making an attempt to use what we have in the Go SDK instead?

Comment (rghetia) [reply receiver/hostmetricsreceiver/collector/cpu/cpu_collector_windows.go:41]: Generating metrics directly like this is fine. If SDK is used then you would need to convert from SDK format to internal format. So from that perspective this is better.

**Comment by james-bebbington** [comment]
See https://github.com/open-telemetry/opentelemetry-collector/pull/848
